epsagon/trace.py

```diff
@@ -46,28 +48,118 @@ def default(self, o):  # pylint: disable=method-hidden
         return output
 
 
+class TraceFactory(object):
+    """
+    A trace factory.
+    """
+
+    def __init__(self):
+        """
+        Initialize.
+        """
+        self.traces = {}
+        self.app_name = ''
+        self.token = ''
+        self.collector_url = ''
+        self.metadata_only = True
+        self.disable_timeout_send = False
+        self.debug = False
+
+    def initialize(
+            self,
+            app_name,
+            token,
+            collector_url,
+            metadata_only,
+            disable_timeout_send,
+            debug
+    ):
+        """
+        Initializes The factory with user's data.
+        User can configure here trace parameters.
+        :param app_name: application name
+        :param token: user's token
+        :param collector_url: the url to send traces to.
+        :param metadata_only: whether to send metadata only or not.
+        :param disable_timeout_send: whether to disable traces send on timeout
+         (when enabled, is done using a signal handler).
+        :param debug: debug flag
+        :return: None
+        """
+
+        self.app_name = app_name
+        self.token = token
+        self.collector_url = collector_url
+        self.metadata_only = metadata_only
+        self.disable_timeout_send = disable_timeout_send
+        self.debug = debug
+
+    def get_trace(self):
+        """
+        Get the trace of the current thread.
+        :return:
+        """
+        thread_id = threading.currentThread().ident
+        if thread_id not in self.traces:
+            new_trace = Trace(
+                self.app_name,
+                self.token,
+                self.collector_url,
+                self.metadata_only,
+                self.disable_timeout_send,
+                self.debug
+            )
+            self.traces[thread_id] = new_trace
```

Comment:
I don't see that you ever remove the trace from this dict - it will cause a memory leak.
You should do that after attempting to send the trace, so `send_traces` should be moved to this class and it will always send the trace of the current thread.

Reply:
Oops, didn't push it :|